python/tk_multi_publish2/api/item.py

```diff
@@ -0,0 +1,1012 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from collections import defaultdict
+
+import inspect
+import os
+import tempfile
+
+import sgtk
+
+from .data import PublishData
+from .task import PublishTask
+
+logger = sgtk.platform.get_logger(__name__)
+
+
+class PublishItem(object):
+    """
+    The ``PublishItem`` class is used to represent something to be published.
+    """
+
+    __slots__ = [
```

Comment:
Since there was a lot of new untested code, I added slots to catch naming errors while assigning a value. I actually caught a couple of bugs in the code due to these, which is awesome.

While this prevents us and clients from doing typos when assigning an attribute (item.visble = True would raise an error right away), it also prevents people from adding their own custom data to our objects. I feel our APIs would be more robust and leaner (using slots implies memory usage doing down quite a bit) if we went this route.

I think people would rather catch bugs easily than be able to assign unrelated data to attributes on an object from the API.

Reply:
__slots__ is cool, but there are some gotchas we'd need to take into account before deciding to use them.

One solvable one is complaints during pickling. Older pickling protocols give you a TypeError:

```python
>>> pickle.loads(pickle.dumps(f))
TypeError: a class that defines __slots__ without defining __getstate__ cannot be pickled
```
That is fixable by using a more recent protocol, but we'd need to make sure we do that with slotted objects:

```python
>>> pickle.loads(pickle.dumps(f, -1))
<__main__.Foo object at 0x1129C770>

>>> pickle.loads(pickle.dumps(f, 2))
<__main__.Foo object at 0x1129C770>
```

Another caveat is that slots reduces your options during inheritance. I had to track this conversation down:
https://mail.python.org/pipermail/python-list/2007-October/432030.html

I'd say we should avoid it if we're providing something we're intending people to subclass.

I'm not sure we'd want to disable __dict__ syntax in general too, since assigning some random var on an object is a useful way to pass data around. The size and speed improvements are very tempting though! 👍

Reply:
Yup, I'm aware of the limitions with pickling, but since we're dumping out stuff to JSON I felt like it was a non-issue.

Inheritance wise, the concerns around multiple inheritance are real, but this is a non-issue in Toolkit and can be easily fixed if it becomes an issue by removing `__slots__` from the class.

I think the author of that post is down-playing the risk of doing mistakes by mistyping and overplays the usefulness of being able to tack-on random data to objects. I can probably count on the fingers of my hand the number of times we've done this in Toolkit.

We can turn back on __dict__ and __weakref__ by adding them to __slots__, so we get the space and speed increase while maintaining flexibility.

As for inheriting objects that have slots, I think this used to be an older behaviour of __slots__. I've experimented with them and if the derived class declares __slots__ it will concatenate wit the base class. I found this [StackOverflow post](https://stackoverflow.com/questions/472000/usage-of-slots) a great summary of the pros and cons of __slots__.

Reply:
Definitely your (and @manneohrstrom's) call. All feels like it depends on the intended use of the code. For hooks where speed probably doesn't matter, random assignment could be more useful, and we're intending them to be subclassed it would seem a little restrictive. For other aspects of TK it feels like it could be a good fit.

Reply:
That's the thing. If the derived class doesn't have a __slots__, you can treat the class as having an implicit value of `[__dict__, __weakref__]` assigned to `__slots__`.

For class that clients have to explicitly derive from, like `Application`, `Engine`, etc, it doesn't make much sense, aside from the space savings (would be nice to benchmark that at some point to see if we should put this into practice).

For classes like the `sgtk.authentication.DefaultsManager` or the `ShotgunModel`, which can optionally be derived from, then it makes sense as most people who use them would benefit from being able to catch assignment errors. If they did derive, then by default they'll get the __dict__ and weakref back.

I'm going to read more on this, but I feel like we could have an easy win here for property heavy APIs.

Reply:
Yeah, I'm removing the user of pickling in the comments. Thanks for pointing it out.